src/core/channels/presence_channel.ts

```diff
@@ -66,15 +73,14 @@ export default class PresenceChannel extends PrivateChannel {
     var data = message.data;
     switch (event) {
       case "pusher_internal:subscription_succeeded":
-        this.subscriptionPending = false;
-        this.subscribed = true;
-        if (this.subscriptionCancelled) {
-          this.pusher.unsubscribe(this.name);
-        } else {
+        if (!this.subscriptionCancelled) {
           this.members.onSubscription(data);
-          this.emit("pusher:subscription_succeeded", this.members);
         }
-        break;
+        this.handleSubscriptionSucceededMessage({
+          event: event,
+          data: this.members,
+        });
```

Comment:
I think that we should do this _before_ calling `members.onSubscription`. That way, we make sure that the internal state is consistent and we know how to proceed if a member does something unexpected (e.g. unsubscribes).

Reply:
I think we need to update the state of `members` before emitting an `event` containing the members!

Reply:
You're right. My comment was actually misleading. Let me try again.

I think that we should update `subscribed` and `subscriptionPending` before we call `members.onSubscription`.

Does this make sense?

Reply:
We could certainly do that. I initially moved this around so that we hanlded subscription state in only one place (the `Channel` class). Partly inspired by your comment about duplication

> This code is a duplicate of `Channel.handleMessage`, except for the call to `members.onSubscription`. Can we reduce the duplication?

Reply:
@hosamaly 1f9fbc70a8306da5d3a0bfa8d368061ba7cc516c this is the alternative imo. I don't think it's terrible to update the membership before updating the internal state, but I suppose this is more 'correct'